### PR TITLE
bug: prevent removing full project by mistake due to build or generate paths

### PR DESCRIPTION
**Pull request by pimlie**
## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

## Description
When you set `options.buildDir` or `options.generate.dir` to a path equal or lower then `options.rootDir`, you will completely remove your project when you build or generate your project as these paths are indiscriminately removed with `fs-extra.remove` in `builder/builder.js` and `builder/generator.js` . I have added a test in options.js which will throw an error when this would be the case.

I feel I need to clarify for everybody here that I did **not** make that mistake while playing around and that because I did **not** make that mistake I did **not** lose my whole project and will **not** have to revert to a day old backup :blush: :cry: 

## Checklist:
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/common/options.js [hunk 1]
```diff
@@ -60,6 +60,25 @@ Options.from = function (_options) {
     : options.rootDir
   options.buildDir = path.resolve(options.rootDir, options.buildDir)
 
+  // Dummy proof buildDir and generate.dir as those folders
+  //  will be completely removed when used
+  if (options.rootDir.startsWith(options.buildDir)) {
+    const errorMessage =
+      'options.buildDir cannot be set lower than or equal to rootDir'
```
Comment (pi0) [inline lib/common/options.js]: `options.buildDir cannot be inside or same as rootDir` isn't better?

File: lib/common/options.js [hunk 2]
```diff
@@ -60,6 +60,25 @@ Options.from = function (_options) {
     : options.rootDir
   options.buildDir = path.resolve(options.rootDir, options.buildDir)
 
+  // Dummy proof buildDir and generate.dir as those folders
+  //  will be completely removed when used
+  if (options.rootDir.startsWith(options.buildDir)) {
```
Comment (pi0) [inline lib/common/options.js]: Not sure. But for more security maybe we have to `path.resolve(options.rootDir)`

Comment (pi0) [reply lib/common/options.js]: Not needed :D Sorry

Comment (pimlie) [reply lib/common/options.js]: Already did on line 44 :)

Comment (pi0) [reply lib/common/options.js]: Saw it too late :)

File: lib/common/options.js (hunk 1 above)
Comment (pimlie) [reply lib/common/options.js]: For me inside means on a higher level, not a lower level. So `/path/subpath` is inside `/path` which is the opposite of what we mean. If you think that as well, we could maybe use outside?

Comment (pi0) [reply lib/common/options.js]: True! I'm drunk man

**Comment by pi0** [comment]
@pimlie If srcDir is different than rootDir isn't better to check it too? I suggest adding a helper function :)

File: lib/common/options.js (hunk 1 above)
Comment (pimlie) [reply lib/common/options.js]: Hmm, while changing this I am thinking that outside is not specific enough. Eg `/build-path` is outside `/nuxt/my-project` and should normally not be a problem. Its really only paths up the same tree as rootDir

**Comment by pimlie** [comment]
Thanks, would probably be best to check srcDir as well ;)

**Comment by pimlie** [comment]
Forgot to add unit tests for the guardDir helper function, see new commit

File: lib/common/utils.js [hunk 3]
```diff
@@ -326,3 +327,26 @@ export const createRoutes = function createRoutes(files, srcDir, pagesDir) {
...
+      (
+        dir1.startsWith(dir2) &&
+        !path.basename(dir1).startsWith(path.basename(dir2))
+      )
+    )
+  ) {
+    const errorMessage =
+      `options.${key2} cannot be a parent of or same as ${key1}`
+    consola.fatal(errorMessage)
+    throw errorMessage
```
Comment (pi0) [inline lib/common/utils.js]: Should be an instance of `Error`: `throw: new Error(errorMessage)`

**Review by pi0 — APPROVED** [review]

**Comment by pi0** [comment]
Thanks @pimlie :)

**Comment by pi0** [comment]
@pimlie That would be a really nice regression test to prevent any unwanted changes. If you want to make this enhancement, I would suggest scanning filesystem once before and once after build step. If any file outside of `buildDir` and `generate.dir` has been changed, then we should fail the tests. I also suggest using [klaw-sync](https://npmjs.org/package/klaw-sync) package for this purpose.

**Pull request merged**